pkg/services/hivelocity/device/device.go

```diff
@@ -168,13 +172,26 @@ func GetFirstFreeDevice(ctx context.Context, hvclient hvclient.Client, machineTy
 		return devices[i].DeviceId < devices[j].DeviceId
 	})
 
-	device := findAvailableDeviceFromList(devices, machineType, hvCluster.Name)
-	return device, nil
+	hvDevices := findAvailableDeviceFromList(devices, hvMachineSpec.DeviceSelector, hvCluster.Name, machineName)
+	if hvDevices == nil {
+		conditions.MarkFalse(
+			hvCluster,
+			infrav1.DeviceAssociateSucceededCondition,
+			infrav1.NoAvailableDeviceReason,
+			clusterv1.ConditionSeverityWarning,
+			"no associated device found",
+		)
+		return nil, nil
+	}
+
+	conditions.MarkTrue(hvCluster, infrav1.DeviceAssociateSucceededCondition)
+	return hvDevices, nil
 }
 
-func findAvailableDeviceFromList(devices []hv.BareMetalDevice, deviceType infrav1.HivelocityDeviceType, clusterName string) *hv.BareMetalDevice {
-	for _, device := range devices {
+func findAvailableDeviceFromList(devices []hv.BareMetalDevice, deviceSelector infrav1.DeviceSelector, clusterName, machineName string) *hv.BareMetalDevice {
```

Comment:
I think you made some mistakes while rebasing.. The machineName shouldn't be here I think